packages/server/src/fhir/operations/utils/agentutils.ts

```diff
@@ -0,0 +1,202 @@
+    try {
+      return await repo.readReference<Device>({ reference: destination });
+    } catch (_err) {
+      return undefined;
+    }
+  }
+  if (destination.startsWith('Device?')) {
+    return repo.searchOne<Device>(parseSearchRequest(destination));
+  }
+  if (contentType === ContentType.PING && (isIPv4(destination) || isValidHostname(destination))) {
+    return { resourceType: 'Device', url: destination };
+  }
+  return undefined;
+}
+
+export async function handleBulkAgentOperation(
+  req: FhirRequest,
+  handler: (agent: Agent) => Promise<FhirResponse>
+): Promise<FhirResponse> {
+  const { repo } = getAuthenticatedContext();
+
+  if (req.query._count && Number.parseInt(req.query._count, 10) > MAX_AGENTS_PER_PAGE) {
+    return [badRequest(`'_count' of ${req.query._count} is greater than max of ${MAX_AGENTS_PER_PAGE}`)];
+  }
+
+  const agents = await getAgentsForRequest(req, repo);
+  if (!agents?.length) {
+    return [badRequest('No agent(s) for given query')];
+  }
+
+  const promises = agents.map((agent: Agent) => handler(agent));
+  const results = await Promise.allSettled(promises);
+  const entries = [] as BundleEntry<Parameters>[];
+
+  for (let i = 0; i < results.length; i++) {
+    const result = results[i];
+    if (result.status === 'rejected') {
+      entries.push(
+        makeResultWrapperEntry(
+          result.reason instanceof OperationOutcomeError ? result.reason.outcome : serverError(result.reason as Error),
+          agents[i]
+        )
+      );
+      continue;
+    }
+    const [outcome, params] = result.value;
+    if (!(isOk(outcome) && params)) {
+      entries.push(makeResultWrapperEntry(outcome, agents[i]));
+      continue;
+    }
+    entries.push(makeResultWrapperEntry(params as Parameters, agents[i]));
+  }
+
+  return [
+    allOk,
+    {
+      resourceType: 'Bundle',
+      type: 'collection',
+      entry: entries,
+    } satisfies Bundle<Parameters>,
+  ];
+}
+
+export function makeResultWrapperEntry(result: Parameters | OperationOutcome, agent: Agent): BundleEntry<Parameters> {
+  return {
+    resource: {
+      resourceType: 'Parameters',
+      parameter: [
+        { name: 'agent', resource: agent },
+        { name: 'result', resource: result },
+      ],
+    },
+  };
+}
+
+export interface AgentMessageOptions {
+  waitForResponse: boolean;
+  timeout?: number;
+}
+
+export async function publishAgentMessage<T extends BaseAgentMessage = BaseAgentMessage>(
+  agent: Agent,
+  message: BaseAgentRequestMessage,
+  options?: AgentMessageOptions
+): Promise<[OperationOutcome] | [OperationOutcome, T | AgentError]> {
+  if (options?.waitForResponse) {
+    let resolve: (response: [OperationOutcome, T | AgentError]) => void;
+    let reject: (err: OperationOutcomeError) => void;
+
+    // If a callback doesn't already exist on the message, tie callback to the associated agent and assign a random ID
+    message.callback ??= getReferenceString(agent) + '-' + randomUUID();
```

Comment:
Do we set `message.callback` anywhere else?  Sorry if I missed it.  Might be nice to only set it exactly once here, so we don't need the conditional logic.